### PR TITLE
Use iterator 4

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

File: inc/ticket.class.php [hunk 1]
```diff
@@ -2372,21 +2320,37 @@ function getActiveOrSolvedLastDaysTicketsForItem($itemtype, $items_id, $days) {
...
+               'NOT' => [
+                  $this->getTable() . '.status' => array_merge(
+                     $this->getClosedStatusArray(),
+                     $this->getSolvedStatusArray()
+                  )
+               ],
+               'AND' => [
+                  'NOT' => [$this->getTable() . '.solvedate' => null],
+                  new \QueryExpression(
+                     "ADDDATE(" . $this->getTable() . ".solvedate, INTERVAL $days DAY) > NOW()"
```
Comment (cedric-anne) [inline inc/ticket.class.php]:
```suggestion
                     "ADDDATE(`" . $this->getTable() . "`.`solvedate`, INTERVAL $days DAY) > NOW()"
```

File: inc/tickettemplatepredefinedfield.class.php [hunk 2]
```diff
@@ -243,127 +243,128 @@ static function showForTicketTemplate(TicketTemplate $tt, $withtemplate = 0) {
...
+      echo "<div class='spaced'>";
+      if ($canedit && $numrows) {
+         Html::openMassiveActionsForm('mass'.__CLASS__.$rand);
+         $massiveactionparams = ['num_displayed' => min($_SESSION['glpilist_limit'], $numrows),
+                                       'container'     => 'mass'.__CLASS__.$rand];
+         Html::showMassiveActions($massiveactionparams);
+      }
+      echo "<table class='tab_cadre_fixehov'>";
+      echo "<tr class='noHover'><th colspan='3'>";
+      echo self::getTypeName($DB->numrows($result));
```
Comment (cedric-anne) [inline inc/tickettemplatepredefinedfield.class.php]:
```suggestion
      echo self::getTypeName($numrows);
```

File: inc/change_ticket.class.php [hunk 3]
```diff
@@ -237,22 +237,33 @@ static function showForChange(Change $change) {
...
+               'ON' => [
+                  'glpi_changes_tickets'  => 'tickets_id',
+                  'glpi_tickets'          => 'id'
+               ]
+            ]
+         ],
+         'WHERE'           => [
+            'glpi_changes_tickets.changes_id'   => $ID
+         ],
+         'ORDERY'          => [
```
Comment (orthagh) [inline inc/change_ticket.class.php]:
```suggestion
         'ORDERBY'          => [
```

File: inc/change_ticket.class.php [hunk 4]
```diff
@@ -343,23 +354,35 @@ static function showForTicket(Ticket $ticket) {
...
+               'ON' => [
+                  'glpi_changes_tickets'  => 'changes_id',
+                  'glpi_changes'          => 'id'
+               ]
+            ]
+         ],
+         'WHERE'           => [
+            'glpi_changes_tickes.tickets_id'   => $ID
+         ],
+         'ORDERY'          => [
```
Comment (orthagh) [inline inc/change_ticket.class.php]:
```suggestion
         'ORDERBY'          => [
```

File: inc/change_ticket.class.php [hunk 5]
```diff
@@ -343,23 +354,35 @@ static function showForTicket(Ticket $ticket) {
...
+         'LEFT JOIN'       => [
+            'glpi_changes' => [
+               'ON' => [
+                  'glpi_changes_tickets'  => 'changes_id',
+                  'glpi_changes'          => 'id'
+               ]
+            ]
+         ],
+         'WHERE'           => [
+            'glpi_changes_tickes.tickets_id'   => $ID
```
Comment (orthagh) [inline inc/change_ticket.class.php]:
```suggestion
            'glpi_changes_tickets.tickets_id'   => $ID
```

File: inc/projecttask_ticket.class.php [hunk 6]
```diff
@@ -370,105 +370,132 @@ static function showForTicket(Ticket $ticket) {
...
+                     'glpi_projecttasks'  => 'projects_id',
+                     'glpi_projects'      => 'id'
+                  ]
+               ]
+            ],
+            'WHERE'     => [
+               'glpi_projecttasks_tickets.tickets_id' => $ID
+            ],
+            'ORDERBY'   => [
+               "$sort $order"
```
Comment (orthagh) [inline inc/projecttask_ticket.class.php]: not here precisely, but $sort seems to contains backquotes, see l371

File: inc/slalevel_ticket.class.php [hunk 7]
```diff
@@ -142,14 +157,32 @@ static function cronSlaTicket(CronTask $task) {
...
-      $query = "SELECT `glpi_slalevels_tickets`.*, `glpi_slas`.`type` as type
-                FROM `glpi_slalevels_tickets`
-                LEFT JOIN `glpi_slalevels`
-                     ON (`glpi_slalevels_tickets`.`slalevels_id` = `glpi_slalevels`.`id`)
-                LEFT JOIN `glpi_slas` ON (`glpi_slalevels`.`slas_id` = `glpi_slas`.`id`)
-                WHERE `glpi_slalevels_tickets`.`date` < NOW()";
+      $iterator = $DB->request([
+         'SELECT'    => [
+            'glpi_slalevels_tickets.*',
+            'glpi_slas.typa AS type',
```
Comment (orthagh) [inline inc/slalevel_ticket.class.php]:
```suggestion
            'glpi_slas.type AS type',
```

**Pull request merged**